### PR TITLE
Add relax_with_penalty

### LINKED ISSUE #3034: Debuggability of JuMP models
One area where I think we can do much more to help users is to provide tooling and documentation to help them debug and test the optimization models that they build.

I had this as one of the items in https://github.com/jump-dev/JuMP.jl/issues/2348, but I'm lifting it to a stand-alone issue.

I think what this needs is a new tutorial that goes over some common issues, and provides links to other resources.

If anyone has ideas/links, please post them below.

## Motivation

@mtanneau's [JuMP-dev talk](https://www.youtube.com/watch?v=61fnHmri7_Q
) discussed a lot of this, but it keeps coming up. I see a lot of common issues on Discourse where people are missing some strategies to debug things.

## Prior art

It's also a topic that is surprisingly lacking in other modeling libraries.
  * GAMS has a nice section: https://www.gams.com/latest/docs/UG_ExecErrPerformance.html. It's probably the most complete I could find.
  * Pyomo doesn't have much https://pyomo.readthedocs.io/en/stable/model_debugging/index.html
  * AMPL has a short section in Chapter 1 on "Catching errors." But it basically advises people to add `check`s to their code to validate data, and that "you may have to spend some time reviewing your formulation before you discover what is wrong." The tooling suggestion is to print out a constraint or objective to look at the data. But that doesn't scale to large problems very well.
  * Gurobi has `feasrelax` and `IIS` support: https://www.gurobi.com/documentation/9.5/examples/diagnose_and_cope_with_inf.html
  * Localsolver has a short section on debugging: https://www.localsolver.com/docs/last/guidelines/debuggingamodel.html

I found this excellent paper that should be required reading for anyone who encounters a bug modelling:

* Pannell, Kingwell, and Schilizzi (1996). Debugging mathematical programming models: principles and practical strategies. _Review of Marketing and Agricultural Economics_, 64(1), 86--100. [pdf](https://github.com/jump-dev/JuMP.jl/files/9255639/64010086.pdf). 

It does have one great paragraph: "Just like science, debugging cannot be a cold, calculating and linear process. Both involve essential elements of inspired guesswork, hutches and sudden flashes of insight which cut through the mist."

They lay out a good framework for debugging as a table/flow-chart. We could aim to replicate that.

Other links

 * https://medium.com/bcggamma/five-strategies-for-debugging-mip-models-4f715809f9f0
 * http://yetanothermathprogrammingconsultant.blogspot.com/2018/08/the-best-way-to-debug-infeasible-models.html

## Documentation: debugging Julia code in general

Give examples of how different common Julia errors can arise (BoundsError, MethodError, ...)

 * Invalid JuMP syntax
 * off-by-one index issues
 * Strategies for debugging
   * Simplify
   * Comment lines
   * Remove macros

Other resources

 * https://benlauwens.github.io/ThinkJulia.jl/latest/book.html#chap21

## Documentation: debugging solver issues

If solver returns something unexpected, 99% of the time there is a bug in your code

  * Test with different solvers
  * Double check the formulation. In most cases, if it gets something different to what you expect, it's because the formulation isn't doing what you think it's doing

Solver returns something unexpected.

 * If a problem is unbounded, what are the likely causes, how to identify which variables are the problem
 * If a problem is infeasible, how to diagnose
 * Numerical issues

Other resources

 * https://www.gurobi.com/documentation/9.5/refman/guidelines_for_numerical_i.html
 * https://www.gams.com/blog/2020/07/dealing-with-models-that-are-unbounded/

Erwin has a good tactic for debugging unbounded problems: http://yetanothermathprogrammingconsultant.blogspot.com/2018/08/the-best-way-to-debug-infeasible-models.html

 * Add the objective as an equality constraint to a new variable with a large bound, then look at the solution.

So 
```julia
model = Model()
@variable(model, x >= 0)
@objective(model, Max, 2x + 1)
```
becomes
```julia
model = Model()
@variable(model, x >= 0)
@variable(model, z  <= 100_000)
@objective(model, Max, z)
@constraint(model, z == 2x + 1)
```

## Documentation: unit testing

 * Write functions not scripts
 * Parameterize your function
 * Test on small inputs
 * Validate those small inputs
 * Use `@test` to ensure
 * Use `primal_feasibility_report` to validate (in)feasible solutions

## Tooling: feasibility relaxation

Some solvers have native support, but we should be able to write this as a function in JuMP.

## Tooling: how to prove your formulation is correct

I've never really seen a good answer for this, and this just stung me. My model was:
```julia
function create_model(neighbors::Dict{Int,Vector{Int}})
    B = 1:500
    T = 1:500
    v(b, t) = (1 + b) * 0.99^t  # faked
    model = Model()
    @variable(model, 0 <= w[B, vcat(0, T)] <= 1)
    @expression(model, y[b = B, t = T], w[b, t] - w[b, t-1])
    @objective(model, Max, sum(v(b, t) * y[b, t] for b in B, t in T))
    @constraints(model, begin
        # I typed
        # [b = setdiff(B, 0), t = T], w[b, t] <= sum(w[b′, t] for b′ in neighbors[b])
        # Should have been (t-1)
        [b = setdiff(B, 0), t = T], w[b, t] <= sum(w[b′, t-1] for b′ in neighbors[b])
        [t = T], sum(y[b, t] for b in B) <= 1
        [b = B, t = T], w[b, t-1] <= w[b, t]
    end)
    set_upper_bound.(w[:, 0], 0.0)
    fix(w[0, 0], 1.0; force = true)
    return model
end
```
So we need some way of easily validating inputs and outputs. We have `primal_feasibility_report`, so perhaps we could have some unit testing framework that let you provide sets of variable values to assert that they are (in)feasible.

@mtanneau pointed out that small test cases in unit tests can be brittle if there are multiple optimal solutions, you're using a local solver, or you're timing out early.

### PR DESCRIPTION
**Pull request by odow**
x-ref https://github.com/jump-dev/MathOptInterface.jl/pull/1995
Closes #3034 

Preview: https://jump.dev/JuMP.jl/previews/PR3140/tutorials/getting_started/debugging/

### CODE REVIEW

File: docs/src/tutorials/getting_started/debugging.jl [hunk 1]
```diff
@@ -230,6 +230,51 @@ termination_status(model)
...
+
+# Here `map` is a dictionary which maps constraint indices to an affine
+# expression representing ``(y + z)``.
+
+# If we optimize the relaxed model, this time we get a feasible solution:
+
+optimize!(model)
+termination_status(model)
+
+# Iterate over the contents of `map` to see which connstraints are violated:
```
Comment (odow) [inline docs/src/tutorials/getting_started/debugging.jl:253]:
```suggestion
# Iterate over the contents of `map` to see which constraints are violated:
```

File: src/constraints.jl [hunk 2]
```diff
@@ -1344,3 +1344,107 @@ function all_constraints(
...
+    before calling [`penalty_relaxation`](@ref).
+
+## Reformulation
+
+See [`MOI.Utilities.ScalarPenaltyRelaxation`](@ref) for details of the
+reformulation.
+
+For each constraint `ci`, the penalty passed to
+[`MOI.Utilities.ScalarPenaltyRelaxation`](@ref) is `get(penalties, ci, default)`.
+If the value is `nothing`, because `ci` does not exist in `penalties` and
```
Comment (blegat) [inline src/constraints.jl:1370]: What if the user puts `nothing` as a value in the dictionary ?

**Comment by blegat** [comment]
According to the JuMP style guide:

> Omit `!` when the name itself makes it clear that modification is taking place, e.g., `add_constraint` and `set_name`

I don't think that we are in the case where "the name itself makes it clear that modification is taking place"

File: src/constraints.jl (hunk 2 above)
Comment (odow) [reply src/constraints.jl:1370]: They can't. I'll update the type signature.

**Comment by odow** [comment]
Updated to `penalty_relaxation!`

**Comment by odow** [comment]
Bike-shedding, how about one of:

 * `make_penalty_relaxation(model)`
 * `create_penalty_relaxation(model)`
 * `add_penalty_relaxation(model)`
 * `reformulate_as_penalty_relaxation(model)`
 * `modify_model_for_penalty_relaxation(model)`

**Comment by odow** [comment]
Following today's call, I've renamed to `relax_with_penalty!`.

**Review by blegat — APPROVED** [review]

**Comment by odow** [comment]
I'll wait a couple more days for comments, and then I'll merge.

**Pull request merged**